### PR TITLE
[compiler] refactor PType.store, SType.coerceOrCopy to take/return values

### PR DESCRIPTION
**Pull request by patrick-schultz**
I didn't give very good names to all the introduced `memoize`s, but those will all go away as we push the refactoring through.

### CODE REVIEW

**Comment by johnc1231** [comment]
@patrick-schultz Lots of test failures

**Comment by patrick-schultz** [comment]
Not surprising, looking

**Comment by patrick-schultz** [comment]
Most of the failures were coming from `SUnreachableValue`s inheriting from both `SValue` and `SCode`, which has caused me problems before too. Since `SCode` will be going away, I just gave in and duplicated them all for now.

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SInsertFieldsStruct.scala [hunk 1]
```diff
@@ -97,10 +97,12 @@ final case class SInsertFieldsStruct(virtualType: TStruct, parent: SBaseStruct,
...
   }
 
-  override def _coerceOrCopy(cb: EmitCodeBuilder, region: Value[Region], value: SCode, deepCopy: Boolean): SCode = {
+  override def _coerceOrCopy(cb: EmitCodeBuilder, region: Value[Region], value: SValue, deepCopy: Boolean): SValue = {
     value match {
-      case ss: SInsertFieldsStructCode if ss.st == this => value
-      case _ => throw new RuntimeException(s"copy insertfields struct")
+      case ss: SInsertFieldsStructValue if ss.st == this => value
+      case _ =>
+        println(s"coerce ${value.st} to ${this}")
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SInsertFieldsStruct.scala:104]: Remove `println`

**Review by johnc1231 — CHANGES_REQUESTED** [review]

**Review by johnc1231 — APPROVED** [review]

**Pull request merged**